XpressAI/xircuits
# ✨ 🐛 Enhancing Copy-Paste Functionality and Keyboard Events

# Description

This PR fixes bugs and enhances the cut-copy-pasting related functions and refines the event handling process. Specifically:

1. Prevent keyboard events from being fired multiple times when having multiple canvas tabs.
2. Enabled spawning on mouse coordinates. 
3. Enabled cross-canvas paste. 
4. Enabled cut-copy-pasting from the keyboard shortcuts.
5. Enabled links and link points for cut-copy-pasting.
6. Updated clone action from happening during the copy process.

## References

https://github.com/XpressAI/xircuits/pull/105
https://github.com/XpressAI/xircuits/issues/184

## Pull Request Type


## Type of Change


# Tests

**1. Test for Spawning on Mouse Coordinates**

    1. Cut / copy a component
    2. Right click on a canvas area and paste spawn
    3. Ensure that the new component appears around at the mouse coordinates.

**2. Test for Cut-Copy-Pasting from the Keyboard**

    1. Select a component and cut or copy using keyboard commands.
    2. Paste using keyboard commands.
    3. Ensure the component is cut-copied-pasted correctly.
    
**3. Test for Keyboard Events on Multiple Canvas Tabs**

    1. Open multiple canvas tabs.
    2. Fire a keyboard event (like cut-copy-paste).
    5. Ensure the event is fired and processed only once.
   
**4. Test for Cross-Canvas Paste**

    1. Copy a component from one canvas.
    2. Switch to a different canvas and paste.
    3. Ensure the component appears in the new canvas.

**5. Test for Cut-Copy-Pasting Links and Link Points**

    1. Create link points on a link and perform a cut or copy operation.
    2. Paste it.
    3. Ensure that the link and link points is cut-copied-pasted correctly.

## Tested on?


# Notes

Currently we're hardcoding the CustomNodeActions in the source code. In the future we should properly implement them as [Jupyterlab configs](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#plugin-settings).
